### PR TITLE
Add a method for replacing terms in the transition relation

### PR DESCRIPTION
**Pull request by makaimann**
This can be used to prune out pieces of a transition system, e.g. by replacing terms with fresh input variables. This is an EXPERTS ONLY feature which makes no guarantee about the resulting system.

FYI @ntsis

### CODE REVIEW

**Comment by makaimann** [comment]
This PR should not be merged until https://github.com/makaimann/smt-switch/pull/130 is merged and the setup script for `smt-switch` is changed back to pulling from the master branch.

**Review by lonsing — APPROVED** [review]
Looks good to me!

This new term replacement feature is a powerful tool. I just left a comment regarding functional transition systems and its invariants, on which the COI implementation relies.

File: core/ts.cpp [hunk 1]
```diff
@@ -597,6 +598,56 @@ bool TransitionSystem::no_next(const Term & term) const
...
+  term_to_name_ = new_term_to_name;
+
+  // NOTE: don't need to update vars, let COI reduction handle that
+  UnorderedTermMap new_state_updates;
+  Term sv;
+  for (auto elem : state_updates_) {
+    sv = elem.first;
+    new_state_updates[sw.visit(sv)] = sw.visit(elem.second);
+  }
+  state_updates_ = new_state_updates;
```
[the remark this answers is not in the document]
Comment (makaimann) [reply core/ts.cpp:648]: Really good point! We should check for that -- I added that in https://github.com/upscale-project/pono/pull/120/commits/52c5c56af65f421b9c79dd9e33c78ee10eaeeab6

**Comment by makaimann** [comment]
Thanks for the review @lonsing!

**Pull request merged**